### PR TITLE
QUIC SRT GEN: Add SRT generator

### PR DESCRIPTION
**Pull request by hlandau**
Trivial little utility that came out of my server refactoring/stateless reset handling renovation. Carving it out since it's easy to review and put in. Harmless if unused so it can't hold up 3.3 in master.

### CODE REVIEW

File: ssl/quic/quic_srt_gen.c
Comment (t8m) [inline ssl/quic/quic_srt_gen.c:19]: Bikeshed: Should we use something more modern such as KMAC with SHAKE-128?

File: ssl/quic/quic_srt_gen.c
Comment (paulidale) [inline ssl/quic/quic_srt_gen.c:70]: I'm a bit uncomfortable using HMAC here.  Over in `reset_token_obfuscate()` I suggested AES with a random key.  This has the advantages of being much faster than HMAC and it is guaranteed to be unique.

File: ssl/quic/quic_srt_gen.c
Comment (paulidale) [reply ssl/quic/quic_srt_gen.c:19]: I suggested AES ECB with random key ...

**Comment by hlandau** [comment]
Worth noting RFC 9000 directly suggests HMAC:

>A single static key can be used across all connections to the same endpoint by generating the proof using a pseudorandom function that takes a static key and the connection ID chosen by the endpoint (see [Section 5.1](https://www.rfc-editor.org/rfc/rfc9000.html#connection-id)) as input. An endpoint could use HMAC [[RFC2104](https://www.rfc-editor.org/rfc/rfc9000.html#RFC2104)] (for example, HMAC(static_key, connection_id)) or the HMAC-based Key Derivation Function (HKDF) [[RFC5869](https://www.rfc-editor.org/rfc/rfc9000.html#RFC5869)] (for example, using the static key as input keying material, with the connection ID as salt). The output of this function is truncated to 16 bytes to produce the stateless reset token for that connection.

I think the performance concern around HMAC is overblown. Willing to consider CMAC if people really care about it.

I don't see how AES can work for the reason given above.

**Comment by paulidale** [comment]
It's a _free_ performance gain on most platforms:

```
type                 16 bytes
hmac(sha256)         90540.25k
cmac(aes-128-cbc)   201773.02k
```

File: ssl/quic/quic_srt_gen.c
Comment (paulidale) [reply ssl/quic/quic_srt_gen.c:19]: KMAC's pretty slow currently and less well tested.

GMAC might be worth consideration: it's just raw AES GCM.

**Comment by hlandau** [comment]
What are you asking for here?

CMAC-AES-128-CBC I am OK with.

I don't see the applicability of AES-128-GCM. Are you proposing we essentially put the input through an AEAD and just take the authentication tag? How do you propose to manage 96-bit IVs statelessly? Unclear to me how this is suitable but maybe I'm missing something.

File: ssl/quic/quic_srt_gen.c
Comment (t8m) [reply ssl/quic/quic_srt_gen.c:19]: GMAC should not be IMO used in this kind of contexts.

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
Since we're down a full-time committer now, I wonder if any of the other @openssl/committers are interested in reviewing this. It's fairly small and fairly pure crypto code.

File: ssl/quic/quic_srt_gen.c
Comment (mattcaswell) [inline ssl/quic/quic_srt_gen.c:68]: Perhaps use SHA256_DIGEST_LENGTH here?

**Comment by hlandau** [comment]
Trivial update to fix build.info conflict after another PR merged.

File: ssl/quic/quic_srt_gen.c
Comment (mattcaswell) [inline ssl/quic/quic_srt_gen.c:81]: Nowhere do we seem to check that `outl >= sizeof(*token)`.

Although strictly shouldn't it look more like this:

````
    memcpy(token->token, mac, sizeof(token->token));
````

File: test/quic_srt_gen_test.c
Comment (mattcaswell) [inline test/quic_srt_gen_test.c:68]: Shouldn't we really pass `token->token` here?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by hlandau** [comment]
Updated.

A QUIC stateless reset token can't be shorter than a SHA256 hash so I think an assert is more than adequate there.

**Review by t8m — APPROVED** [review]
Still OK

**Review by mattcaswell — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master. Thanks for the reviews.

**Pull request closed without merging**
